core/src/main/java/org/frankframework/pipes/Json2XmlValidator.java

```diff
@@ -186,60 +187,148 @@ private static double getQuality(String mimeType) {
+			}
+		} else if (inputFormat == DocumentFormat.JSON) {
 			if (!isAllowJson() && !responseMode) {
-				return getErrorResult("message is not XML, because it starts with ["+firstChar+"] and not with '<'", session, responseMode);
+				return getErrorResult("message is not XML, because it starts with [" + findFirstChar(input) + "] and not with '<'", session, responseMode);
 			}
-			if (firstChar!='{' && firstChar!='[') {
-				return getErrorResult("message is not XML or JSON, because it starts with ["+firstChar+"] and not with '<', '{' or '['", session, responseMode);
+		} else if (!Message.isEmpty(input)) {
+			return getErrorResult("message is not XML or JSON, because it starts with [" +findFirstChar(input)+ "] and not with '<', '{' or '['", session, responseMode);
+		}
+
+		Message messageToValidate;
+		if (inputFormat == null) {
+			messageToValidate = new Message("{}");
+			storeInputFormat(getOutputFormat(), input, session, responseMode); //Message is empty, but could be either XML or JSON. Look at the accept header, and if not set fall back to the default OutputFormat.
+		} else {
+			if (Message.isNull(input)) {
+				messageToValidate = new Message("{}");
+			} else {
+				messageToValidate = input;
 			}
 			storeInputFormat(DocumentFormat.JSON, input, session, responseMode);
 		}
-
 		try {
 			return alignJson(messageToValidate, session, responseMode);
 		} catch (XmlValidatorException e) {
 			throw new PipeRunException(this, "Cannot align JSON", e);
 		}
 	}
 
+	/**
+	 * Peek into the message to find the first character, and from that determine the input format.
+	 * For reasons of backwards compatibility:
+	 * <ul>
+	 *     <li> If the message is a null-message ({@link Message#isNull()}, then the input format is determined to be JSON and the actual message parsed will be {@code "{}"}.</li>
+	 *     <li> If the message is an empty message ({@link Message#isEmpty()}, then the input format is {@code null}, but still the actual message parsed will be {@code "{}"}.</li>
+	 * </ul>
+	 * Otherwise, the determination of input format is as follows:
+	 * <ul>
+	 *     <li> If the first non-whitespace character is an {@code '<'} the input format is XML</li>
+	 *     <li> If the first non-whitespace character is a {@code '{'}} or {@code '['} then the input format is JSON</li>
+	 *     <li> If there was no non-whitespace character in the first 1KiB of the message, or the first non-whitespace character was a different character, return {@code null}.</li>
+	 * </ul>
+	 *
+	 * @param input Message from which to read.
+	 * @return DocumentFormat of the input message, if it was possible to determine, or null.
+	 * @throws IOException Exception from reading the message.
+	 */
+	private @Nullable DocumentFormat findDocumentFormat(Message input) throws IOException {
+		if (Message.isNull(input)) {
+			return DocumentFormat.JSON;
+		}
+		if (Message.isEmpty(input)) {
+			return null;
+		}
+		Reader reader = input.asReader();
```

Comment:
Afhankelijk van de onderliggende code kan dit wel een behoorlijke buffer zijn. Ik vind het stiekem mooier om toch de magic te gebruiken 😞

Reply:
Ik zal er eens naar kijken, maar ik denk dat dezelfde onderliggende buffering heeft dus ik weet niet of het veel uitmaakt.